pulsar-broker/src/test/java/org/apache/pulsar/broker/service/KeySharedE2ETest.java

```diff
@@ -0,0 +1,242 @@
+
+        AtomicInteger trigCount = new AtomicInteger();
+
+        public TestConsumerStateEventListener(String name) {
+            this.name = name;
+        }
+
+        @Override
+        public void becameActive(Consumer<?> consumer, int partitionId) {
+        }
+
+        @Override
+        public void becameInactive(Consumer<?> consumer, int partitionId) {
+
+        }
+
+        @Override
+        public void keySharedRuleChanged(Consumer<?> consumer, Predicate<String> keyPredicate) {
+            this.keyPredicate = keyPredicate;
+            trigCount.incrementAndGet();
+        }
+    }
+
+    protected void clearEventListener(TestConsumerStateEventListener eventListener){
+        eventListener.keyPredicate = null;
+        eventListener.trigCount.set(0);
+    }
+
+    @Test
+    public void testConsumerEventsWithAutoHashRangeImpl() throws Exception {
+        final String topicName = "persistent://prop/use/ns-abc/key-shared-topic_01-" + System.currentTimeMillis();
+        final String subName = "sub_key_shared_01";
+        ConsumerBuilder<byte[]> consumerBuilder = pulsarClient.newConsumer().topic(topicName).subscriptionName(subName)
+                .subscriptionType(SubscriptionType.Key_Shared);
+        int rangeSize = 0;
+        // 1.One subscription.
+        TestConsumerStateEventListener listener1 = new TestConsumerStateEventListener("key_shared_listener_01");
+        ConsumerBuilder<byte[]> consumerBuilder1 = consumerBuilder.clone().consumerName("key_shared_consumer_01")
+                .consumerEventListener(listener1);
+        Consumer<byte[]> consumer1 = consumerBuilder1.subscribe();
+        Thread.sleep(CONSUMER_ADD_OR_REMOVE_WAIT_TIME);
+        Assert.assertEquals(listener1.trigCount.get(), 1);
+        Assert.assertTrue(listener1.keyPredicate != null);
+        Assert.assertTrue(listener1.keyPredicate instanceof Predicate4HashRangeAutoSplitStickyKeyConsumerSelector);
+        Predicate4HashRangeAutoSplitStickyKeyConsumerSelector predicate4ConsistentHashingStickyKeyConsumerSelector1 =
+                (Predicate4HashRangeAutoSplitStickyKeyConsumerSelector) listener1.keyPredicate;
+        rangeSize = predicate4ConsistentHashingStickyKeyConsumerSelector1.getRangeSize();
+        Assert.assertTrue(predicate4ConsistentHashingStickyKeyConsumerSelector1.getRangeSize()
+                == predicate4ConsistentHashingStickyKeyConsumerSelector1.getHighHash());
+        Assert.assertEquals(predicate4ConsistentHashingStickyKeyConsumerSelector1.getLowHash(), -1);
+        // 2.Two subscription.
+        TestConsumerStateEventListener listener2 = new TestConsumerStateEventListener("key_shared_listener_02");
+        ConsumerBuilder<byte[]> consumerBuilder2 = consumerBuilder.clone().consumerName("key_shared_consumer_02")
+                .consumerEventListener(listener2);
+        Consumer<byte[]> consumer2 = consumerBuilder2.subscribe();
+        Thread.sleep(CONSUMER_ADD_OR_REMOVE_WAIT_TIME);
```

Comment:
use `Awaitility.await()` instead of Thread.sleep

Reply:
Good idea ! Thanks for your guidance.  I'm so sorry that I'm so busy these days